File: SPCode/Data/Services/SPDataTracer.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports System.Data
Imports System.Diagnostics

Namespace Data.Services
    Friend Class SPDataTracer
        Private Shared ReadOnly tracerSwitch As TraceSwitch

        Private Sub New()
        End Sub
        Shared Sub New()
            tracerSwitch = New TraceSwitch("TraceSwitch", "application trace switch")
        End Sub

        Public Shared Sub Write(ByVal cmd As IDbCommand)
            If tracerSwitch.Level >= TraceLevel.Verbose Then
                Trace.WriteLine(GetCommandDescription(cmd))
            End If
        End Sub

        Public Shared Sub Write(ByVal err As Exception)
            If tracerSwitch.Level >= TraceLevel.Error Then
                Trace.WriteLine(GetExceptionDescription(err))
            End If
        End Sub

        Private Shared Function GetCommandDescription(ByVal cmd As IDbCommand) As String
            Dim sb As StringBuilder = New StringBuilder(300)

            sb.Append(Constants.vbCrLf)
            sb.Append("-- CommandText: " & Constants.vbCrLf)
            sb.Append(cmd.CommandText & Constants.vbCrLf)
            sb.Append("-- CommandType: " & Constants.vbCrLf)
            sb.Append(cmd.CommandType.ToString() & Constants.vbCrLf)
            sb.Append("-- Parameters: " & Constants.vbCrLf)

            Dim i As Integer = 0
            Do While i < cmd.Parameters.Count
                Dim parmater As IDataParameter = CType(cmd.Parameters(i), IDataParameter)
                sb.Append(i & ". Name: " & parmater.ParameterName & ", Value: " & parmater.Value.ToString & ", Type: " & parmater.DbType & ", Direction: " & parmater.Direction & "." & Constants.vbCrLf)
                i += 1
            Loop

            sb.Append(Constants.vbCrLf)
            Return sb.ToString()
        End Function

        Private Shared Function GetExceptionDescription(ByVal e As Exception) As String
            Dim sb As StringBuilder = New StringBuilder(200)

            sb.Append("Error: " & e.Message).Append(Constants.vbCrLf)
            sb.Append("Type: " & e.GetType().FullName).Append(Constants.vbCrLf)
            sb.Append("Source: " & e.Source).Append(Constants.vbCrLf)
            sb.Append("Target: " & e.TargetSite.ToString).Append(Constants.vbCrLf)
            sb.Append("Stack: " & e.StackTrace).Append(Constants.vbCrLf)

            e = e.InnerException
            Do While Not e Is Nothing
                sb.Append(" ---------- Inner Exception: ---------- ").Append(Constants.vbCrLf)
                sb.Append(GetExceptionDescription(e))
                e = e.InnerException
            Loop

            Return sb.ToString()
        End Function
    End Class
End Namespace

File: SPCode/CS/Data/Services/SPDataTracer.cs
using System.Text.RegularExpressions;
using System.Diagnostics;
using System;
using System.Drawing.Drawing2D;
using System.Windows.Forms;
using System.Collections;
using System.Drawing;
using Microsoft.VisualBasic;
using System.Data.SqlClient;
using System.Data;
using System.Collections.Generic;
using WeifenLuo.WinFormsUI;
using Microsoft.Win32;
using WeifenLuo;
using System.Text;


namespace SoftLogik.Win
{
	namespace Data
	{
		namespace Services
		{
			
			internal class SPDataTracer
			{
				
				private static readonly TraceSwitch tracerSwitch;
				
				private SPDataTracer()
				{
				}
				static SPDataTracer()
				{
					tracerSwitch = new TraceSwitch("TraceSwitch", "application trace switch");
				}
				
				public static void Write(IDbCommand cmd)
				{
					if (tracerSwitch.Level >= TraceLevel.Verbose)
					{
						Trace.WriteLine(GetCommandDescription(cmd));
					}
				}
				
				public static void Write(System.Exception err)
				{
					if (tracerSwitch.Level >= TraceLevel.Error)
					{
						Trace.WriteLine(GetExceptionDescription(err));
					}
				}
				
				private static string GetCommandDescription(IDbCommand cmd)
				{
					StringBuilder sb = new StringBuilder(300);
					
					sb.Append(Constants.vbCrLf);
					sb.Append("-- CommandText: " + Constants.vbCrLf);
					sb.Append(cmd.CommandText + Constants.vbCrLf);
					sb.Append("-- CommandType: " + Constants.vbCrLf);
					sb.Append(cmd.CommandType.ToString() + Constants.vbCrLf);
					sb.Append("-- Parameters: " + Constants.vbCrLf);
					
					int i = 0;
					while (i < cmd.Parameters.Count)
					{
						IDataParameter parmater = (IDataParameter) (cmd.Parameters[i]);
						sb.Append(i + ". Name: " + parmater.ParameterName + ", Value: " + parmater.Value.ToString() + ", Type: " + parmater.DbType + ", Direction: " + parmater.Direction + "." + Constants.vbCrLf);
						i++;
					}
					
					sb.Append(Constants.vbCrLf);
					return sb.ToString();
				}
				
				private static string GetExceptionDescription(Exception e)
				{
					StringBuilder sb = new StringBuilder(200);
					
					sb.Append("Error: " + e.Message).Append(Constants.vbCrLf);
					sb.Append("Type: " + e.GetType().FullName).Append(Constants.vbCrLf);
					sb.Append("Source: " + e.Source).Append(Constants.vbCrLf);
					sb.Append("Target: " + e.TargetSite.ToString()).Append(Constants.vbCrLf);
					sb.Append("Stack: " + e.StackTrace).Append(Constants.vbCrLf);
					
					e = e.InnerException;
					while (e != null)
					{
						sb.Append(" ---------- Inner Exception: ---------- ").Append(Constants.vbCrLf);
						sb.Append(GetExceptionDescription(e));
						e = e.InnerException;
					}
					
					return sb.ToString();
				}
			}
		}
	}
	
	
}
